Hookable menus

deivid-rodriguez opened the pull request:
#### :tophat: What? Why?

Fixes #131.

The first commit is what I need for the [assemblies feature](https://github.com/deivid-rodriguez/decidim/commit/c4ba1a8d6cfeeac5ea1ed6a0750ad01da42a4e32) so I don't have to monkeypatch any views in the plugin. Instead, I can add an initializer like this:

```ruby
initializer "decidim_assemblies.menu" do
  Decidim.menu :menu do |menu|
    menu.item -> { I18n.t("menu.assemblies", scope: "decidim") },
              -> { decidim_assemblies.assemblies_path },
              position: 2.5,
              active: :inclusive
  end
end
```

I was continuing to complete all the specifications in #131 (https://github.com/deivid-rodriguez/decidim/commit/c4ba1a8d6cfeeac5ea1ed6a0750ad01da42a4e32), but I decided to stop and grab some feedback first.
 
Also, I haven't looked closely but maybe this also helps with integrating the `decidim-accountability` plugin?

#### :pushpin: Related Issues
- Fixes #131.
- Related to #1346.
- Related to decidim/decidim-accountability#4.

#### :clipboard: Subtasks
- [x] Basic menus.
- [x] Add conditional display parameter.
- [x] Add items with icons (<strike>separate `icon_item` DSL method?,</strike> conditional behavior based on the presence of an `icon_name` parameter?).
- [x] Migrate system & admin menus.

### :camera: Screenshots (optional)
_None_.

#### :ghost: GIF
![chaplin_telas](https://cloud.githubusercontent.com/assets/2887858/26424565/0c50ae28-40a8-11e7-9354-c72d20f1476b.gif)

josepjaume commented:
Hi @deivid-rodriguez, this looks really nice!

Using lambdas is quite succint. I don't have a special preference about it, but in decidim we tend to use APIs like:

```ruby
initializer "decidim_assemblies.menu" do
  Decidim.menu :menu do |menu|
    menu.item do |item|
      item.title I18n.t("menu.assemblies", scope: "decidim")
      item.path decidim_assemblies.assemblies_path
      item.position 2.5
      item.active :inclusive
    end
end
```

It feels a bit more expressive, but maybe it's me. Also, with this API, you can skip the lambdas because `|item|` can get evaluated in the context of the renderer - you just store the block that resolves into `|item|` and evaluate it later.

This technique can also be used in your approach though, with a bit of extra work.

deivid-rodriguez commented:
@josepjaume Thanks for the feedback!

I like the API you initially proposed in https://github.com/decidim/decidim-accountability/issues/4#issuecomment-300175718 so I tried to follow it here. I do admit to it's a bit uglier with the lambdas and your new idea is cleaner in that regard.

I still like it because:

* It resembles to the underlying `link_to` form, so it should be easy to grasp for Rails devs. I mean, two compulsory parameters, the first one being the label and the second one the url, and then a trailing hash of extra optional parameters.

* It would easily allow to directly pass any extra unknown parameters to the underlying `link_to` without having to extend the API. For example for setting extra CSS classes for specific elements or things like that. Not sure how easy that would be in the alternative proposal. Maybe through some `method_missing` trickery, but if feels nicer to just have a hash of parameters in case we wanted to implement something like that.

So I'd like to keep this form if you don't have a strong opinion.

Also, admin menus come with a bit of extra icon candy (https://github.com/decidim/decidim/blob/master/decidim-admin/app/helpers/decidim/admin/application_helper.rb#L18-L24), we should probably shape the API to account for that. I thought of a separate API method `icon_item`, or an extra `icon` parameter to the `item` method, not sure if you have any other ideas or preferences here.

josepjaume commented:
Okay then! I like it. 

But maybe you could skil the lambdas if you built the menu on rendering time? Just store all the menu block and evaluate them afterwards. Does that sound like a bad idea? Would clean it a bit for sure, but maybe we'd be missing some opportunities...?

Also about the icon candy, not really sure about it. Would just specifying classes be enough?

deivid-rodriguez commented:
> But maybe you could skil the lambdas if you built the menu on rendering time? Just store all the menu block and evaluate them afterwards. Does that sound like a bad idea? Would clean it a bit for sure, but maybe we'd be missing some opportunities...?

Yeah, I'd also like to clean up the lambdas. I'm not sure if you would ever want to specifically evaluate the items at initialization time, can't think of something like that right now so your idea make sense to me... I'll experiment a bit more and see if I can put it into code.

> Also about the icon candy, not really sure about it. Would just specifying classes be enough?

Mmmm not sure what you mean, like specifying a CSS class and showing the icon via CSS instead of via markup?

deivid-rodriguez commented:
@josepjaume I pushed some changes here, now the API looks as you designed it! :smiley: 

On a side note, I'm having problems to get some PR's built, see https://circleci.com/gh/decidim/decidim/1238 for example). It seems that the circle config is not correctly picked up (https://discuss.circleci.com/t/circleci-not-always-picking-up-config-yml/11921, maybe?). I wonder if this has to do with opening PR's from a forked external repo...

I see that I'm in the developers group of decidim, does that mean I have push access to https://github.com/decidim/decidim?

beagleknight commented:
Hi @deivid-rodriguez !
I'm not sure if it's related to these options:

![image](https://cloud.githubusercontent.com/assets/106021/26484943/267a63ea-41f5-11e7-828c-6d5a781572b3.png)

I think we may need to enable the second option but I am not really sure.

deivid-rodriguez commented:
@beagleknight Maybe, I'm not sure either. It's fine though, it does not happen very often :)

josepjaume commented:
This is so awesome! Can't wait to see it hooked to the entire system!

deivid-rodriguez commented:
Ok, I'm pretty much done with this. Tomorrow I'll give "extending menus" a bit of testing but you can hit me with your feedback!

Copying here how the admin menu is now registered to show how the API looks like (basically what @josepjaume proposed):

```ruby
initializer "decidim_admin.menu" do
  Decidim.menu :admin_menu do |menu|
    menu.item I18n.t("menu.dashboard", scope: "decidim.admin"),
              decidim_admin.root_path,
              icon_name: "dashboard",
              active: ["decidim/admin/dashboard" => :show]

    menu.item I18n.t("menu.participatory_processes", scope: "decidim.admin"),
              decidim_admin.participatory_processes_path,
              icon_name: "target",
              active: :inclusive

    menu.item I18n.t("menu.participatory_process_groups", scope: "decidim.admin"),
              decidim_admin.participatory_process_groups_path,
              icon_name: "layers",
              active: :inclusive

    menu.item I18n.t("menu.static_pages", scope: "decidim.admin"),
              decidim_admin.static_pages_path,
              icon_name: "book",
              active: :inclusive,
              if: can?(:read, Decidim::StaticPage)

    menu.item I18n.t("menu.users", scope: "decidim.admin"),
              decidim_admin.users_path,
              icon_name: "person",
              active: [%w(decidim/admin/user_groups decidim/admin/users), []],
              if: can?(:read, :admin_users)

    menu.item I18n.t("menu.newsletters", scope: "decidim.admin"),
              decidim_admin.newsletters_path,
              icon_name: "envelope-closed",
              active: :inclusive,
              if: can?(:index, Decidim::Newsletter)

    menu.item I18n.t("menu.settings", scope: "decidim.admin"),
              decidim_admin.edit_organization_path,
              icon_name: "wrench",
              active: [%w(decidim/admin/organization decidim/admin/scopes), []],
              if: can?(:read, current_organization)
  end
end
```

josepjaume commented:
Do you consider this finished? Should we merge? I'm planning a release and this'd be a nice feature to have.

deivid-rodriguez commented:
Hei! I wanted to try extending the admin menu and also fix some docs that got outdated after refactoring. I'll call this ready in a bit!

josepjaume commented:
Awesome! Can't wait!

josepjaume commented:
And BTW @deivid-rodriguez  I just made you collaborator of the project ;)

deivid-rodriguez commented:
Ok, so I tested this, fixed a few issues I found and updated docs. Tests started failing for me locally after all of today's dance in master, I probably need to start fresh. But I think it's unrelated to this PR since they also fail against my local master branch.

mrcasals commented:
@deivid-rodriguez but.. but... CircleCI is green... 😮

deivid-rodriguez commented:
> @deivid-rodriguez but.. but... CircleCI is green... :open_mouth:

It was #1416. Now things are more similar in both environments, so weird things like this will be less likely to happen!

> Amazing!!

Thanks for your kind words! :heart:
